monitoring-dashboard/components/org.wso2.ei.dashboard.core/src/main/java/org/wso2/ei/dashboard/core/rest/delegates/heartbeat/HeartBeatDelegate.java

```diff
@@ -80,63 +83,80 @@ private boolean isNodeRegistered(HeartbeatObject heartbeat) {
-                heartbeatScheduledExecutorService.shutdownNow();
-                deleteNode(productName, heartbeat);
-            }
-        };
-        heartbeatScheduledExecutorService.scheduleWithFixedDelay(runnableTask, 3 *
-                heartbeatInterval, 3 * heartbeatInterval, TimeUnit.SECONDS);
-        return dataManager.insertHeartbeat(heartbeat, accessToken);
+                boolean isNodeDeregistered =
+                        (currentTimestamp - longTimestampOfRegisteredNode) > 3 * heartbeatInterval * 1000;
+                if (isNodeDeregistered) {
+                    logger.info("Node : " + heartbeat.getNodeId() + " of group : " + heartbeat.getGroupId() + " has " +
+                            "de-registered. Hence deleting node information");
+                    heartbeatScheduledExecutorService.shutdownNow();
+                    deleteNode(productName, heartbeat);
+                }
+            };
+            heartbeatScheduledExecutorService.scheduleWithFixedDelay(runnableTask, 3 *
+                    heartbeatInterval, 3 * heartbeatInterval, TimeUnit.SECONDS);
+            return dataManager.insertHeartbeat(heartbeat, accessToken);
+        } catch (DashboardServerException e) {
+            throw new ManagementApiException (e.getMessage(), 500);
+        }
     }
 
     private boolean isNodeShutDown(HeartbeatObject heartbeat, String initialTimestamp) {
         return !dataManager.checkIfTimestampExceedsInitial(heartbeat, initialTimestamp);
     }
 
     private void deleteNode(String productName, HeartbeatObject heartbeat) {
-        int rowCount = dataManager.deleteHeartbeat(heartbeat);
-        if (rowCount > 0) {
-            logger.info("Successfully deleted node where group_id : " + heartbeat.getGroupId() + " and node_id : "
-                     + heartbeat.getNodeId() + ".");
-            deleteAllNodeData(productName, heartbeat);
-        } else {
+        try {
+            int rowCount = dataManager.deleteHeartbeat(heartbeat);
+            if (rowCount > 0) {
+                logger.info("Successfully deleted node where group_id : " + heartbeat.getGroupId() + " and node_id : "
+                        + heartbeat.getNodeId() + ".");
+                deleteAllNodeData(productName, heartbeat);
+            } else {
+                throw new DashboardServerException("Error occurred while deleting node where group_id : "
+                        + heartbeat.getGroupId() + " and node_id : " + heartbeat.getNodeId()
+                        + ".");
+            }
+        } catch (DashboardServerException e) {
             throw new DashboardServerException("Error occurred while deleting node where group_id : "
-                                               + heartbeat.getGroupId() + " and node_id : " + heartbeat.getNodeId()
-                                               + ".");
+                    + heartbeat.getGroupId() + " and node_id : " + heartbeat.getNodeId() + ".");
```

Comment:
With this the exception stack trace will be lost?

Reply:
Added e.getCause() now.